basket/base/middleware.py

```diff
@@ -0,0 +1,58 @@
+import contextlib
+import inspect
+import time
+
+from django.conf import settings
+from django.http import Http404
+from django.utils.deprecation import MiddlewareMixin
+
+from basket import metrics
+
+
+class HostnameMiddleware(MiddlewareMixin):
+    """Add header with k8s cluster / pod details for debugging."""
+
+    def process_response(self, request, response):
+        response["X-Backend-Server"] = "/".join(filter(None, [getattr(settings, x) for x in ["CLUSTER_NAME", "K8S_NAMESPACE", "K8S_POD_NAME"]]))
+        return response
+
+
+class MetricsStatusMiddleware(MiddlewareMixin):
+    """Send status code counts to statsd"""
+
+    def process_response(self, request, response):
+        metrics.incr(f"response.{response.status_code}")
+        return response
+
+    def process_exception(self, request, exception):
+        if not isinstance(exception, Http404):
+            metrics.incr("response.500")
+
+
+class MetricsRequestTimingMiddleware(MiddlewareMixin):
+    """Send request timing to statsd"""
+
+    def process_view(self, request, view_func, view_args, view_kwargs):
+        if inspect.isfunction(view_func):
+            view = view_func
+        else:
+            view = view.__class__
+
+        with contextlib.suppress(AttributeError):
+            request._start_time = time.time()
+            request._view_module = view.__module__
+            request._view_name = view.__name__
+
+    def _record_timing(self, request):
+        if hasattr(request, "_start_time"):
+            view_time = int((time.time() - request._start_time) * 1000)
+            metrics.timing(f"view.{request._view_module}.{request._view_name}.{request.method}", view_time)
+            metrics.timing(f"view.{request._view_module}.{request.method}", view_time)
+            metrics.timing(f"view.{request.method}", view_time)
+
+    def process_response(self, request, response):
+        self._record_timing(request)
```

Comment:
In a future change, you should add a tag with the HTTP status code here and in `process_exception` so then you can differentiate timings between HTTP 200s and the various error situations which will take less time and muck your statistics.

Reply:
neat!

Reply:
Something like this?
```python
metrics.timing(f"view.{request.method}", view_time, tags=[generate_tag("status_code", status_code)])
```

Reply:
Yes! Also, since the tag and value are pretty benign, you could shorten it to:

```python
metrics.timing(f"view.{request.method}", view_time, tags=[f"status_code:{status_code}"])
```